### PR TITLE
build: use Micronaut Framework 4.0.0-SNAPSHOT

### PR DESCRIPTION
**Pull request by sdelamo**

### CODE REVIEW

**Comment by sdelamo** [comment]
@dstepanov maybe you can help me out. 

We have to modify the code to use the new API introduced here https://github.com/micronaut-projects/micronaut-core/pull/8217 

```java
./gradlew :function-aws-api-proxy:compileJava  
Type-safe project accessors is an incubating feature.
Project accessors enabled, but root project name not explicitly set for 'buildSrc'. Checking out the project in different folders will impact the generated code and implicitly the buildscript classpath, breaking caching.

> Task :function-aws-api-proxy:compileJava
warning: unknown enum constant When.MAYBE
  reason: class file for javax.annotation.meta.When not found
Note: Creating bean classes for 7 type elements
warning: unknown enum constant When.MAYBE
  reason: class file for javax.annotation.meta.When not found
warning: unknown enum constant When.MAYBE
  reason: class file for javax.annotation.meta.When not found
warning: unknown enum constant When.MAYBE
  reason: class file for javax.annotation.meta.When not found
/Users/sdelamo/github/micronaut-projects/micronaut-aws/function-aws-api-proxy/src/main/java/io/micronaut/function/aws/proxy/MicronautAwsProxyResponse.java:228: error: MicronautAwsProxyResponse.AwsHeaders is not abstract and does not override abstract method setConversionService(ConversionService) in ConversionServiceAware
    private class AwsHeaders implements MutableHttpHeaders {
            ^
/Users/sdelamo/github/micronaut-projects/micronaut-aws/function-aws-api-proxy/src/main/java/io/micronaut/function/aws/proxy/MicronautLambdaContainerHandler.java:360: error: no suitable method found for executeRoute(MicronautAwsProxyRequest<CAP#1>,boolean,Flux<RouteMatch<?>>)
            routeResponse = routeExecutor.executeRoute(
                                         ^
    method RouteExecutor.executeRoute(RequestBodyReader,HttpRequest<?>,boolean,StaticResourceResponseFinder) is not applicable
      (actual and formal argument lists differ in length)
    method RouteExecutor.executeRoute(AtomicReference<HttpRequest<?>>,boolean,boolean,ExecutionFlow<RouteMatch<?>>) is not applicable
      (actual and formal argument lists differ in length)
  where CAP#1 is a fresh type-variable:
    CAP#1 extends Object from capture of ?
/Users/sdelamo/github/micronaut-projects/micronaut-aws/function-aws-api-proxy/src/main/java/io/micronaut/function/aws/proxy/MicronautLambdaContainerHandler.java:366: error: incompatible types: ExecutionFlow<MutableHttpResponse<?>> cannot be converted to Supplier<ExecutionFlow<MutableHttpResponse<?>>>
            routeResponse = Flux.from(routeE`xecutor.filterPublisher(new AtomicReference<>(request), routeExecutor.onError(e, request)));
                                                                                                                         ^
/Users/sdelamo/github/micronaut-projects/micronaut-aws/function-aws-api-proxy/src/main/java/io/micronaut/function/aws/proxy/MicronautLambdaContainerHandler.java:710: error: incompatible types: Publisher<MutableHttpResponse<?>> cannot be converted to Supplier<ExecutionFlow<MutableHttpResponse<?>>>
        Flux.from(routeExecutor.filterPublisher(requestReference, responsePublisher))
```

Moreover, `AwsHeaders` seems to need to implement [`ConversionServiceAware.java `](https://github.com/micronaut-projects/micronaut-core/blob/4.0.x/core/src/main/java/io/micronaut/core/convert/ConversionServiceAware.java)


Also, this fails: 

```
./gradlew :function-client-aws:compileTestGroovy
Type-safe project accessors is an incubating feature.
Project accessors enabled, but root project name not explicitly set for 'buildSrc'. Checking out the project in different folders will impact the generated code and implicitly the buildscript classpath, breaking caching.

> Task :function-client-aws:compileTestGroovy
Note: Creating bean classes for 4 type elements
warning: Implicitly compiled files were not subject to annotation processing.
  Use -implicit to specify a policy for implicit compilation.
1 warning

> Task :function-client-aws:compileTestGroovy FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':function-client-aws:compileTestGroovy'.
> io/micronaut/ast/groovy/utils/AstAnnotationUtils
```

@dstepanov It seems you removed AstAnnotationUtils here: https://github.com/micronaut-projects/micronaut-core/commit/752a22117bf80695899b721257e9436292ee3a77#diff-12d26803929b8b3d13e678348c0a42048673cc40649cc7bdda4c41b88e70e5aa

**Comment by timyates** [comment]
Investigating the [failure with Cloudwatch logging](https://ge.micronaut.io/s/rwh5i3pbx7zqa/tests/:aws-cloudwatch-logging:test/io.micronaut.aws.cloudwatch.logging.CloudwatchLoggingSpec/test%20Cloudwatch%20logging?top-execution=1)

**Comment by timyates** [comment]
@sdelamo Ready for review 👍

File: settings.gradle.kts [hunk 1]
```diff
@@ -6,7 +6,7 @@ pluginManagement {
 }
 
 plugins {
-    id("io.micronaut.build.shared.settings") version "5.4.2"
+    id("io.micronaut.build.shared.settings") version "6.1.1"
```
Comment (sdelamo) [inline settings.gradle.kts:9]:
```suggestion
    id("io.micronaut.build.shared.settings") version "6.2.1"
```

File: settings.gradle.kts [hunk 2]
```diff
@@ -36,19 +36,16 @@ include("aws-apigateway")
 include("function-aws-test")
 include("test-suite")
 include("test-suite-aws-sdk-v2")
-include("test-suite-http-server-tck-function-aws-api-proxy")
+//include("test-suite-http-server-tck-function-aws-api-proxy")
```
Comment (sdelamo) [inline settings.gradle.kts:39]: why?

Comment (sdelamo) [reply settings.gradle.kts]: https://github.com/micronaut-projects/micronaut-aws/pull/1499/commits/d900c8c5bf2063203f7ac7f0dbcfcd8d554b6f8f

**Comment by timyates** [comment]
Fixed the TCK build script, we seem to have a couple of TCK failures ion the ErrorHandlerTest, I will investigate

File: aws-distributed-configuration/build.gradle.kts [hunk 3]
```diff
@@ -4,6 +4,6 @@ plugins {
 
 dependencies {
     api(projects.awsCommon)
-    api(mn.micronaut.discovery)
+    api(libs.micronaut.discovery.client)
```
Comment (timyates) [inline aws-distributed-configuration/build.gradle.kts:7]: Once we have a snapshot of https://github.com/micronaut-projects/micronaut-discovery-client/pull/407 we can change this to use the BOM

**Review by timyates — APPROVED** [review]

File: aws-distributed-configuration/build.gradle.kts (hunk 3 above)
Comment (timyates) [reply aws-distributed-configuration/build.gradle.kts]: Ok, we have a snapshot...  updated... lets watch CI

**Comment by timyates** [comment]
@sdelamo 👍

**Pull request merged**